injector.py

```diff
@@ -555,21 +554,24 @@ class SingletonScope(Scope):
     >>> injector = Injector()
     >>> provider = ClassProvider(A)
     >>> singleton = SingletonScope(injector)
-    >>> a = singleton.get(A, provider)
-    >>> b = singleton.get(A, provider)
+    >>> a = singleton.get(injector, BindingKey(A), provider)
+    >>> b = singleton.get(injector, BindingKey(A), provider)
     >>> a is b
     True
     """
     def configure(self):
-        self._context = {}
+        self._context = defaultdict(dict)
 
     @synchronized(lock)
-    def get(self, key, provider):
+    def get(self, injector, key, provider):
+        # Try to find the associated binding for a key, but if that doesn't work,
+        # use the binder the SingletonScope is bound to.
+        binder = injector._get_binder_for(key) or self.injector.binder
         try:
-            return self._context[key]
+            return self._context[key][binder]
         except KeyError:
             provider = InstanceProvider(provider.get(self.injector))
-            self._context[key] = provider
+            self._context[key][self.injector.binder] = provider
```

Comment:
I'm a little bit worried that this may result in an unbounded growth of the context dictionary (if someone creates and quickly discards child injectors repeatedly)